### PR TITLE
[hold] upgrade: Webpack 5

### PR DESCRIPTION
**Pull request by icirellik**
A quick upgrade to Webpack 5 brings with it a savings of ~100 KB gzipped. Along with all the other features of [Webpack 5](https://webpack.js.org/blog/2020-10-10-webpack-5-release/):

* Better Tree-Shaking
* Better Caching
* Module Federation

Before:

webpack-4

After:

webpack-5

### CODE REVIEW

**Comment by damassi** [comment]
@icirellik What does `R` and `U` mean?

**Comment by icirellik** [comment]
Reminders of follow-up work.

File: webpack/envs/clientCommonConfig.js [hunk 1]
```diff
@@ -75,20 +77,14 @@ export const clientCommonConfig = {
...
-      // bundles. This /may/ lead to a broken build when e.g. a reaction
-      // module that's used on the client side imports something from
-      // graphql-js, but that's better than silently including this.
-      new webpack.IgnorePlugin(/^graphql(\/.*)?$/),
-    ],
-    new webpack.NamedModulesPlugin(),
+    // TODO: Why would these end up in the client bundle?
+    new webpack.IgnorePlugin({
+      resourceRegExp: /^react-relay-network-modern-ssr\/node8\/server/,
+    }),
```
Comment (damassi) [inline webpack/envs/clientCommonConfig.js:86]: I think this was added due to how `Artsy/Router/buildServer|ClientApp` used to be exported from reaction, but we import these things directly everywhere now so i think this is safe to remove.

File: webpack/envs/clientCommonConfig.js [hunk 2]
```diff
@@ -135,6 +131,11 @@ export const clientCommonConfig = {
     // Symlink issues should be fixed via `yarn --pnp`
     modules: [path.resolve(basePath, "src"), "node_modules"],
     symlinks: false,
+    fallback: {
+      "path": false,
+      "os": require.resolve("os-browserify/browser"),
+      "buffer": require.resolve("buffer/"),
```
Comment (damassi) [inline webpack/envs/clientCommonConfig.js:137]: What does this section + individual keys do?

File: webpack/envs/clientDevelopmentConfig.js [hunk 3]
```diff
@@ -56,29 +52,13 @@ export const clientDevelopmentConfig = {
...
-      excludeWarnings: true,
-      skipFirstNotification: true,
-    }),
-    new FriendlyErrorsWebpackPlugin({
-      clearConsole: false,
-      compilationSuccessInfo: {
-        messages: [`[Force] Listening on http://localhost:${env.port} \n`],
-        notes: [""],
-      },
-    }),
```
Comment (damassi) [inline webpack/envs/clientDevelopmentConfig.js:81]: Why remove our errors?

File: webpack/envs/clientDevelopmentConfig.js [hunk 4]
```diff
@@ -56,29 +52,13 @@ export const clientDevelopmentConfig = {
...
     new ForkTsCheckerWebpackPlugin({
       formatter: "codeframe",
       formatterOptions: "highlightCode",
       checkSyntacticErrors: true,
       watch: ["./src"],
     }),
-    new ForkTsCheckerNotifierWebpackPlugin({
-      excludeWarnings: true,
-      skipFirstNotification: true,
-    }),
```
Comment (damassi) [inline webpack/envs/clientDevelopmentConfig.js:74]: Why remove notifier? This is helpful

File: webpack/envs/clientDevelopmentConfig.js [hunk 5]
```diff
@@ -56,29 +52,13 @@ export const clientDevelopmentConfig = {
     ],
   },
   plugins: [
-    new CaseSensitivePathsPlugin(),
```
Comment (damassi) [inline webpack/envs/clientDevelopmentConfig.js:59]: This was recently added because it helped us track down  a very pesky case sensitive bug. Why remove?

Comment (damassi) [reply webpack/envs/clientDevelopmentConfig.js:59]: If all of the following removals are due to plugin compat, we should comment them out with a 
```
TODO: Reenable once webpack 5 is supported
```

File: webpack/envs/serverConfig.js [hunk 6]
```diff
@@ -62,18 +62,16 @@ export const serverConfig = {
     minimize: env.isProduction && !env.webpackDebug,
     minimizer: [
       new TerserPlugin({
-        cache: false,
```
Comment (damassi) [inline webpack/envs/serverConfig.js:65]: (nonblocking) This significantly speeds up local prod builds, can we perhaps disable it if we're in a deploy environment?

**Review by damassi — APPROVED** [review]
Thanks for taking this on @icirellik 👍 

A couple nonblocking q's / requests around commenting out incompatable webpack plugins so that we can remember to check up and reenable when ready for webpack 5 (versus deleting outright), but this pushes things forward nicely in the meantime.

File: webpack/envs/clientCommonConfig.js (hunk 2 above)
Comment (icirellik) [reply webpack/envs/clientCommonConfig.js:137]: Webpack 5 no longer provides automatic browser polyfills. This section specifies the recommended polyfill to use for each node api we access in the browser.

File: webpack/envs/serverConfig.js (hunk 6 above)
Comment (icirellik) [reply webpack/envs/serverConfig.js:65]: It's literally no longer an option and has been integrated with their new caching system.

File: webpack/envs/clientDevelopmentConfig.js (hunk 5 above)
Comment (icirellik) [reply webpack/envs/clientDevelopmentConfig.js:59]: Would you explain the particular bug in more detail or provide a link to it? I am intimately familiar with Apples case-insensitive file system and the many issues that it creates in particular when renaming files in git. However git has options to address this already and wonder if we could instead leverage those.

Comment (dzucconi) [reply webpack/envs/clientDevelopmentConfig.js:59]: https://github.com/artsy/force/pull/6462/ was the issue that prompted this — I had downcased a folder name accidently against convention, but imports would work locally with alternate casing. I'd very much like to be warned about this, regardless of where it happens.

Comment (dzucconi) [reply webpack/envs/clientDevelopmentConfig.js:59]: (Er, that's not the issue but rather the fix)

File: package.json [hunk 7]
```diff
@@ -219,23 +219,22 @@
...
-    "@babel/plugin-proposal-optional-chaining": "7.11.0",
-    "@babel/plugin-syntax-dynamic-import": "7.8.3",
-    "@babel/plugin-transform-modules-commonjs": "7.10.4",
-    "@babel/plugin-transform-runtime": "7.11.0",
-    "@babel/preset-env": "7.11.0",
-    "@babel/preset-react": "7.10.4",
-    "@babel/preset-typescript": "7.10.4",
-    "@babel/register": "7.10.5",
-    "@babel/runtime": "7.11.2",
+    "@babel/cli": "^7.12.1",
```
Comment (eessex) [inline package.json:222]: Might be a good idea to extract these babel updates into their own PR if we can easily isolate.

**Review by eessex — APPROVED** [review]

**Comment by icirellik** [comment]
Waiting on: https://github.com/gregberge/loadable-components/issues/636

**Comment by icirellik** [comment]
Will create a new PR when `@loadable` is compatible with Webpack 5 async loading as this PR has become quite stale.
